src/components/utils/include/utils/timer.h

```diff
@@ -84,18 +89,137 @@ class Timer {
    */
   bool IsRunning() const;
   /**
+   * @brief Suspends timer execution after next loop.
+   */
+  virtual void Suspend();
+
+  /**
+   * @brief Stop timer update timeout and start timer again
+   * Note that it cancel thread of timer, If you use it from callback,
+   * it probably will stop execution of callback function
+   * @param timeout_milliseconds new timeout value
+   *
+   */
+  virtual void UpdateTimeOut(const uint32_t timeout_milliseconds);
+
+  /**
    * @brief GetTimeout
    * @return returns timeout
    */
   Milliseconds GetTimeout() const;
 
  private:
+  /**
+   * @brief Delegate release timer, will call callback function one time
+   */
+  class TimerDelegate : public threads::ThreadDelegate {
+    friend class Timer;
+
+   public:
+    /**
+     * @brief Default constructor
+     *
+     * @param timer_thread The Timer_thread pointer
+     */
+    explicit TimerDelegate(Timer* timer);
+
+    /**
+     * @brief Destructor
+     */
+    virtual ~TimerDelegate();
+
+    /**
+     * @brief Thread main function.
+     */
+    void threadMain() OVERRIDE;
+
+    /**
+     * @brief Called by thread::thread to free all allocated resources.
+     */
+    void exitThreadMain() OVERRIDE;
+
+    /**
+     * @brief Set new Timeout
+     * @param timeout_milliseconds New timeout to be set
+     */
+    virtual void SetTimeOut(const uint32_t timeout_milliseconds);
+
+    /**
+      * @brief Wait until timer will start
+      */
+    virtual void WaitUntilStart();
+
+   protected:
+    Timer* timer_;
```

Comment:
@LuxoftAKutsan , it's just pointer to main class that you create when want use timer. So, timer's owner responsible to this pointer.